### PR TITLE
Milestone 6: Friendship v2

### PR DESCRIPTION
**Pull request by rukundoeric**
## What does this do?

- Friendship v2

## Description of Task to be completed?
Implement friendships with 2 rows for mutual friendship.
 As a guest user:
 - I want to be able to send a friendship invitation.
 - I want to see a button “Add Friend” next to the name of user who is not my friend yet - on both users’ list and single user page.
 - I want to be able to see pending friendship invitations sent to me from other users.
 - I want to be able to accept or reject friendship's invitation.
## How should this be manually tested?
- If you don't have ruby or rails installed on your pc, you can follow this [Tuto](https://www.tutorialspoint.com/ruby-on-rails/rails-installation.htm) to do so.
- Clone this repo
- Cd into the clone folder
- run `bundle install` to install the required gems
- then run `yarn install` to setup webpack and install all its packages
- If the installation is complete, go ahead and set up the database by running migration with this command `rails db:migrate`
- Then finally start the app by running `rails s` and then run the application on this URL `http://localhost:3000/` in your browser

### CODE REVIEW

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Great job so far @team, you are almost there you need to make just some minor changes :hammer: :wrench: to get your project approved :checkered_flag: 
check the comment below :point_down: :point_down: :point_down: :point_down: :point_down:

## Required Changes:

- [x] when friendship is confirmed, an additional row with reversed user-friend pair should be created, so when you try to get the list of friends on both sides it's gonna work, what you did is correct you are halfway, but if created a relationship with two users and you open `rails console`, try to get the list of friends from both users will not work only one is working. 

I suggest you use this code to confirm the relationship in the friendship Model:

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] in user models while you were trying to get the pending friends you fished the pending friendships which are wrong to fix this, change your code with this or adapted to it:

```ruby
    # User model
   
     has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
     has_many :pending_friends, through: :pending_friendships, source: :friend

     has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id" 
     has_many :friend_requests, through: :inverted_friendships

```

- [x] A user should only be able to see his own posts and his friends. in your case, any user can see all the posts.
image

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_feel free to contact me on slack if you have any questions._

File: app/models/user.rb [hunk 1]
```diff
@@ -16,8 +16,11 @@ class User < ApplicationRecord
...
   has_many :friend_requests_sent, -> { where(status: 'pending') },
            class_name: 'Friendship', foreign_key: 'user_id'
+  has_many :pending_friends, through: :friend_requests_sent, source: :user
   has_many :friend_requests_received, -> { where(status: 'pending') },
            class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :friend_requests, through: :friend_requests_received, source: :user
+
   def friend?(user)
     friendship = Friendship.find_by(user_id: user.id, friend_id: id, status: 'confirmed') ||
                  Friendship.find_by(user_id: id, friend_id: user.id, status: 'confirmed')
```
Comment (ghost) [inline app/models/user.rb]: Add tests for the above associations and methods. If the methods are not being used outside the class, declare them as private and don't write tests for them.

**Note: I am not able to access lines 15 and above on this file so I will describe the corrections below.**

-  write test for `has_many :likes, dependent: :destroy`, which is on line 11.

- For the `has_many :comments, dependent: :destroy` test, you have omitted the` :dependent` option. Please fix that. Check and fix more sections like this, where options are being omitted,  especially in your `friendship_spec.rb file`.

File: app/models/friendship.rb [hunk 2]
```diff
@@ -3,4 +3,13 @@ class Friendship < ApplicationRecord
...
   enum status: { pending: 0, confirmed: 1 }
+
+  def confirm_friend
+    Friendship.create(user_id: friend_id, friend_id: user_id, status: 'confirmed')
+  end
+
+  after_destroy do |f|
+    friendship = Friendship.find_by(user_id: f.friend_id, friend_id: f.user_id)
+    friendship&.destroy
+  end
```
Comment (ghost) [inline app/models/friendship.rb:14]: Write tests for the above methods or declare them private if you are not using them outside the class.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team 😄,

**Great job implementing the requested changes 💪. However, there are still a few changes needed.**

1). **I will use an example. Say I have two user's Jen and Ken. They are not friends yet. However, when Jen visits Ken's profile, she can see Ken's post below the `add friend` button. Please fix that.**

2).**Using the same example above. In this case, Jen sends Ken a friend request. When I check Jen's `pending_friends`, I get Jen as the pending friend instead of Ken.**

3). **Use the correct format for writing your spec tests. If you are writing tests for your models then they should be inside a `spec/models ` directory.**

4). **Please write tests for your `views` and `helpers`. Also, you omitted tests for the following files in your models: `comment.rb`, `like.rb`, `post.rb`.**

5). **Check more in the comments below.**

## Happy coding 🥂

**PS: You can reach me on slack (Rose Wanjohi).**

**Comment by rukundoeric** [comment]
The project doesn't require to implement real-time functionality, so please refresh your browser before you check if friend-request is received or not.

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team  ,  you have done a brilliant job with this project. :clap:
#### You can merge !!!

**Pull request merged**